python/src/spark_rapids_ml/dbscan.py

```diff
@@ -0,0 +1,655 @@
+
+    def setMinSamples(self: P, value: int) -> P:
+        return self._set_params(min_samples=value)
+
+    def getMinSamples(self) -> int:
+        return self.getOrDefault("min_samples")
+
+    def setMetric(self: P, value: str) -> P:
+        return self._set_params(metric=value)
+
+    def getMetric(self) -> str:
+        return self.getOrDefault("metric")
+
+    def setMaxMbytesPerBatch(self: P, value: Optional[int]) -> P:
+        return self._set_params(max_mbytes_per_batch=value)
+
+    def getMaxMbytesPerBatch(self) -> Optional[int]:
+        return self.getOrDefault("max_mbytes_per_batch")
+
+    def setCalcCoreSampleIndices(self: P, value: bool) -> P:
+        return self._set_params(calc_core_sample_indices=value)
+
+    def getCalcCoreSampleIndices(self) -> bool:
+        return self.getOrDefault("calc_core_sample_indices")
+
+    def _fit(self, dataset: DataFrame) -> _CumlModel:
+        if self.getMetric() == "precomputed":
+            raise ValueError(
+                "Spark Rapids ML does not support the 'precomputed' mode from sklearn and cuML, please use those libraries instead"
+            )
+
+        # Create parameter-copied model without assess the input dataframe
+        # All information will be retrieved from Model and transform
+        model = DBSCANModel(verbose=self.verbose, n_cols=0, dtype="")
+
+        model._num_workers = self.num_workers
+        self._copyValues(model)
+
+        return model
+
+    def _create_pyspark_model(self, result: Row) -> _CumlModel:
+        raise NotImplementedError("DBSCAN does not support model creation from Row")
+
+    def _get_cuml_fit_func(
+        self,
+        dataset: DataFrame,
+        extra_params: Optional[List[Dict[str, Any]]] = None,
+    ) -> Callable[
+        [FitInputType, Dict[str, Any]],
+        Dict[str, Any],
+    ]:
+        raise NotImplementedError("DBSCAN does not can not fit and generate model")
+
+    def _out_schema(self) -> Union[StructType, str]:
+        return StructType()
+
+
+class DBSCANModel(
+    DBSCANClass, _CumlCaller, _CumlModelWithPredictionCol, _DBSCANCumlParams
+):
+    def __init__(
+        self,
+        n_cols: int,
+        dtype: str,
+        verbose: int | bool,
+    ):
+        super(DBSCANClass, self).__init__()
+        super(_CumlModelWithPredictionCol, self).__init__(n_cols=n_cols, dtype=dtype)
+        super(_DBSCANCumlParams, self).__init__()
+
+        self._setDefault(
+            idCol=alias.row_number,
+        )
+
+        self.verbose = verbose
+        self.BROADCAST_LIMIT = 8 << 30
+        self._dbscan_spark_model = None
+
+    def _pre_process_data(self, dataset: DataFrame) -> Tuple[  # type: ignore
+        List[Column],
+        Optional[List[str]],
+        int,
+        Union[Type[FloatType], Type[DoubleType]],
+    ]:
+        (
+            select_cols,
+            multi_col_names,
+            dimension,
+            feature_type,
+        ) = _CumlCaller._pre_process_data(self, dataset)
```

Comment:
This needs to specify the preprocess for Caller since the DBSCANModel is both a caller and a model, both of which contains a pre_process_data method (different ones)